graph-builder/src/registry.rs

```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+fn trim_protocol(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(credentials_path: &Option<PathBuf>) -> Box<dyn std::io::Read> {
+    let empty = Box::new(std::io::empty());
+    match credentials_path {
+        Some(path) => match File::open(path) {
+            Ok(file) => Box::new(file),
+            _ => empty,
+        },
+        _ => empty,
+    }
+}
+
+pub fn authenticate_client<'a>(
+    client: &'a mut dkregistry::v2::Client,
+    login_scope: &'a str,
+) -> impl futures::future::Future<Item = &'a dkregistry::v2::Client, Error = dkregistry::errors::Error>
+{
+    client
+        .is_v2_supported()
+        .and_then(move |v2_supported| {
+            if !v2_supported {
+                Err("API v2 not supported".into())
+            } else {
+                Ok(client)
+            }
+        })
+        .and_then(|dclient| {
+            dclient.is_auth(None).and_then(move |is_auth| {
+                if is_auth {
+                    Err("no login performed, but already authenticated".into())
+                } else {
+                    Ok(dclient)
+                }
+            })
+        })
+        .and_then(move |dclient| {
+            dclient.login(&[&login_scope]).and_then(move |token| {
+                dclient
+                    .is_auth(Some(token.token()))
+                    .and_then(move |is_auth| {
+                        if !is_auth {
+                            Err("login failed".into())
+                        } else {
+                            println!("logged in!");
+                            Ok(dclient.set_token(Some(token.token())))
+                        }
+                    })
+            })
+        })
+}
+
 /// Fetches a vector of all release metadata from the given repository, hosted on the given
 /// registry.
-pub fn fetch_releases(registry: &str, repo: &str) -> Result<Vec<Release>, Error> {
-    let mut metadata = Vec::new();
-    for tag in fetch_tags(registry, repo)? {
-        metadata.push(Release {
-            source: format!(
-                "{}/{}:{}",
-                registry
-                    .trim_left_matches("https://")
-                    .trim_left_matches("http://"),
-                repo,
-                tag
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_path: &Option<PathBuf>,
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_path))
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let login_scope = format!("repository:{}:pull", &repo);
```

Comment:
I prefer not to use intermediate variables unless it helps with readability. In this case, I don't think it buys you much so I would probably inline this.

Reply:
You can discuss this with the compiler

```
error[E0597]: borrowed value does not live long enough                                                                                       
   --> graph-builder/src/registry.rs:125:14                                                                                                  
    |                                                                                                                                        
125 |             &format!("repository:{}:pull", &repo),                                                                                     
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ temporary value does not live long enough                                            
126 |         ))                                                                                                                             
127 |         .map_err(|e| format_err!("{}", e))?;                                                                                           
    |                                            - temporary value dropped here while still borrowed                                         
...                                                                                                                                          
208 | }                                                                                                                                      
    | - temporary value needs to live until here                                                                                             
    |                                                                                                                                        
    = note: consider using a `let` binding to increase its lifetime                                                                          
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
                                                                                                                                             
error: aborting due to previous error                                                                                                        
                                                                                                                                             
For more information about this error, try `rustc --explain E0597`.                                                                          
error: Could not compile `graph-builder`.  
```